### PR TITLE
[Feature] Basic support of encryption in cloud native

### LINKED ISSUE #46223: [Feature] Transparent Data Encryption
- [x] https://github.com/StarRocks/starrocks/pull/47840
- [x] https://github.com/StarRocks/starrocks/pull/46224
- [x] https://github.com/StarRocks/starrocks/pull/52179

## Feature request

Support  transparent data encryption

### PR DESCRIPTION
**Pull request by decster**
## Why I'm doing:

Basic support of encryption in cloud native

## What I'm doing:

* Add encryption support in fs
* Add basic plain key types and key wrap/unwrap
* Add KeyManager in FE and KeyCache in BE
* support encryption for rowset files and cloud native primary index(other data file types support later)

Fixes #46223

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by kevincai — COMMENTED** [review]
openssl is not a multi-thread safe library, calling the library in multi-thread env, may need additional tuning to make it thread-safe.

**Comment by decster** [comment]
> openssl is not a multi-thread safe library, calling the library in multi-thread env, may need additional tuning to make it thread-safe.

which part are you talking about? As for file encryption/decryption, each encrypt/decryption call uses its own context, no shared state involved.

**Comment by kevincai** [comment]
> > openssl is not a multi-thread safe library, calling the library in multi-thread env, may need additional tuning to make it thread-safe.
> 
> which part are you talking about? As for file encryption/decryption, each encrypt/decryption call uses its own context, no shared state involved.

the EVP context creation may be suffered as well.

**Comment by decster** [comment]
> > > openssl is not a multi-thread safe library, calling the library in multi-thread env, may need additional tuning to make it thread-safe.
> > 
> > 
> > which part are you talking about? As for file encryption/decryption, each encrypt/decryption call uses its own context, no shared state involved.
> 
> the EVP context creation may be suffered as well.

    OPENSSL_RET_NOT_OK(EVP_EncryptInit_ex(ctx, cipher, nullptr, eh.key_bytes(), iv), "Failed to initialize encryption");

this call does not modify any shared object, from the openssl documentation, the use is thread safe?

**Comment by kevincai** [comment]
> EVP_CIPHER_CTX_new

will check and get back to you.

**Comment by kevincai** [comment]
> > > > openssl is not a multi-thread safe library, calling the library in multi-thread env, may need additional tuning to make it thread-safe.
> > > 
> > > 
> > > which part are you talking about? As for file encryption/decryption, each encrypt/decryption call uses its own context, no shared state involved.
> > 
> > 
> > the EVP context creation may be suffered as well.
> 
> ```
> OPENSSL_RET_NOT_OK(EVP_EncryptInit_ex(ctx, cipher, nullptr, eh.key_bytes(), iv), "Failed to initialize encryption");
> ```
> 
> this call does not modify any shared object, from the openssl documentation, the use is thread safe?
> 
> or are you sugguesting EVP_CIPHER_CTX_new is not threadsafe and need a global lock to protect?

Forget about it, looks like the newer version of openssl doesn't have such trouble any more. the `CRYPTO_*_id_callback` and `CRYPTO_*_dynlock_*_callback` are all dummy macros, which means useless at all.

File: be/src/agent/heartbeat_server.cpp [hunk 1]
```diff
@@ -74,11 +74,15 @@ void HeartbeatServer::init_cluster_id_or_die() {
 
 void HeartbeatServer::heartbeat(THeartbeatResult& heartbeat_result, const TMasterInfo& master_info) {
     //print heartbeat in every minute
-    LOG_EVERY_N(INFO, 12) << "get heartbeat from FE."
-                          << "host:" << master_info.network_address.hostname
+    LOG_EVERY_N(INFO, 12) << "get heartbeat from FE. host:" << master_info.network_address.hostname
                           << ", port:" << master_info.network_address.port << ", cluster id:" << master_info.cluster_id
                           << ", run_mode:" << master_info.run_mode << ", counter:" << google::COUNTER;
 
+    if (master_info.encrypted != config::encrypted) {
```
Comment (wyb) [inline be/src/agent/heartbeat_server.cpp:81]: FE does not set `master_info.encrypted`

File: be/src/fs/fs_memory.cpp [hunk 2]
```diff
@@ -417,13 +419,15 @@ MemoryFileSystem::~MemoryFileSystem() {
 StatusOr<std::unique_ptr<SequentialFile>> MemoryFileSystem::new_sequential_file(const SequentialFileOptions& opts,
                                                                                 const std::string& path) {
     (void)opts;
```
Comment (wyb) [inline be/src/fs/fs_memory.cpp:421]: remove

File: gensrc/proto/encryption.proto [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+enum EncryptionAlgorithmPB {
+    NO_ENCRYPTION = 0;
+    AES_128 = 1;
+}
+
+message EncryptionKeyPB {
+    optional int64 id = 1; // not required, e.g. DEK may not have an id
+    optional int64 parent_id = 2; // not required, e.g. root key may not have parent
+    optional int64 createTime = 3; // create timestamp
```
[the remark this answers is not in the document]
Comment (decster) [reply gensrc/proto/encryption.proto:31]: fixed

File: be/src/fs/fs_memory.cpp (hunk 2 above)
Comment (decster) [reply be/src/fs/fs_memory.cpp]: fixed

File: fe/fe-core/src/main/java/com/starrocks/encryption/EncryptionKey.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    public boolean isKEK() {
+        return parent != null && parent.isMasterKey();
+    }
+
+    public abstract EncryptionKey generateKey();
+
+    public abstract void decryptKey(EncryptionKey key);
+
+    public void toPB(EncryptionKeyPB pb, KeyMgr mgr) {
```
[the remark this answers is not in the document]
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/encryption/EncryptionKey.java:59]: should be used in subclass

**Review by wyb — DISMISSED** [review]

**Review by wyb — DISMISSED** [review]

**Review by wyb — DISMISSED** [review]

**Review by nshangyiming — DISMISSED** [review]

File: be/src/common/config.h [hunk 5]
```diff
@@ -240,6 +240,7 @@ CONF_mInt32(disk_stat_monitor_interval, "5");
 CONF_mInt32(profile_report_interval, "30");
 CONF_mInt32(unused_rowset_monitor_interval, "30");
 CONF_String(storage_root_path, "${STARROCKS_HOME}/storage");
+CONF_Bool(encrypted, "false");
```
Comment (trueeyu) [inline be/src/common/config.h:243]: Is the naming of this configuration too simple and I can’t tell what is being encrypted?

Comment (decster) [reply be/src/common/config.h:243]: by requirement, all user data on disk should be encrypted, currently including segment/delete files, pk index, datacache,  will add binlog/fulltext index in following PR, probably will add spill output in future.

Comment (decster) [reply be/src/common/config.h:243]: how about enable_transparent_data_encryption?

File: be/src/common/daemon.cpp [hunk 6]
```diff
@@ -298,6 +299,7 @@ void Daemon::init(bool as_cn, const std::vector<StorePath>& paths) {
     LOG(INFO) << DiskInfo::debug_string();
     LOG(INFO) << MemInfo::debug_string();
     LOG(INFO) << base::CPU::instance()->debug_string();
+    LOG(INFO) << "openssl aesni support: " << openssl_supports_aesni();
```
Comment (trueeyu) [inline be/src/common/daemon.cpp:302]: C++ bool type may be printed as invisible characters

**Review by trueeyu — APPROVED** [review]

**Review by nshangyiming — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Comment by decster** [comment]
@mergify backport branch-3.3

**Pull request merged**